### PR TITLE
Better error handling in pipeline.py

### PR DESCRIPTION
**Pull request by szalpal**
#### Why we need this PR?
*Pick one, remove the rest*
- It fixes a bug *bug description*
- It adds new feature needed because of *why we need this feature*
- Refactoring to improve error handling. This is especially useful in dali_backend pipelines, where the common pattern used is:

```python
@piepline_def(...)
def pipe():
    ...
    return x

def main(filename):
    pipe().serialize(filename)
```
Without the change, the code above would explode with meaningless error, saying that you can't use `define_graph` and `set_outputs` at the same time...

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     Added `if` statement
 - Affected modules and functionalities:
     `pipeline.py`
 - Key points relevant for the review:
     NA
 - Validation and testing:
     NA
 - Documentation (including examples):
     NA

**JIRA TASK**: *[Use DALI-XXXX or NA]*

Signed-off-by: Michał Szołucha <mszolucha@nvidia.com>

### CODE REVIEW

**Comment by szalpal** [comment]
!build

File: dali/python/nvidia/dali/pipeline.py [hunk 1]
```diff
@@ -1018,6 +1018,11 @@ def serialize(self, define_graph=None, filename=None):
         kwargs : dict
                 Refer to Pipeline constructor for full list of arguments.
         """
+        if define_graph is not None and not callable(define_graph):
```
Comment (JanuszL) [inline dali/python/nvidia/dali/pipeline.py:1021]: Could you add a test for that?
I'm not sure what is the exact use case where it should help.

File: dali/python/nvidia/dali/pipeline.py [hunk 2]
```diff
@@ -1018,6 +1018,11 @@ def serialize(self, define_graph=None, filename=None):
...
+            raise TypeError(
+                "Provided `define_graph` argument is not callable." +
+                " Didn't you want to write `.serialize(filename=...)`?"
+                if isinstance(define_graph, str) else "")
```
Comment (mzient) [inline dali/python/nvidia/dali/pipeline.py:1025]:
```suggestion
                (" Didn't you want to write `.serialize(filename=...)`?"
                if isinstance(define_graph, str) else ""))
```

**Review by mzient — APPROVED** [review]

File: dali/python/nvidia/dali/pipeline.py (hunk 1 above)
Comment (JanuszL) [reply dali/python/nvidia/dali/pipeline.py:1021]: Also another question is if we should add similar guarding code to other methods?

Comment (szalpal) [reply dali/python/nvidia/dali/pipeline.py:1021]: Sure, added a test.

As for the use case, I've explained it in the PR description, here's a copy-paste:

This is especially useful in dali_backend pipelines, where the common pattern used is:
```python
@piepline_def(...)
def pipe():
    ...
    return x

def main(filename):
    pipe().serialize(filename)
```
Without the change, the code above would explode with meaningless error, saying that you can't use define_graph and set_outputs at the same time...

File: dali/python/nvidia/dali/pipeline.py (hunk 2 above)
Comment (szalpal) [reply dali/python/nvidia/dali/pipeline.py]: Done.

File: dali/python/nvidia/dali/pipeline.py (hunk 1 above)
Comment (JanuszL) [reply dali/python/nvidia/dali/pipeline.py:1021]: I think you should write it guards from the case where the filename is passed as the first positional argument, while it is the second.
This description tells nothing and I have spend a couple of minutes trying to understand the nature of the problem.

**Review by JanuszL — APPROVED** [review]

**Comment by szalpal** [comment]
!build

**Pull request merged**